src/codegen/function_body.rs

```diff
@@ -29,21 +30,64 @@ impl Builder {
         self.from_glib_suffix = prefix_suffix.1;
         self
     }
-    pub fn parameter(&mut self, parameter: String) -> &mut Builder {
-        self.parameters.push(parameter);
+    pub fn parameter(&mut self, parameter: String, out_as_return: bool) -> &mut Builder {
+        self.parameters.push((parameter, out_as_return));
+        if out_as_return { self.outs_as_return = true }
         self
     }
     pub fn generate(&self) -> Vec<String> {
-        let mut v: Vec<String> = Vec::new();
+        let mut v: Vec<String> = Vec::with_capacity(16);
         let unsafed = self.generate_unsafed();
+        if self.outs_as_return {
+            self.write_out_variables(&mut v);
+        }
         write_to_vec!(v, "unsafe {{");
         write_to_vec!(v, "{}{}", tabs(1), unsafed);
         write_to_vec!(v, "}}");
+        if self.outs_as_return {
+            write_to_vec!(v, "{}", self.generate_out_return());
+        }
         v
     }
     fn generate_unsafed(&self) -> String {
-        let param_str = self.parameters.connect(", ");
+        let param_str = self.generate_func_parameters();
         format!("{}ffi::{}({}){}", self.from_glib_prefix,
             self.glib_name, param_str, self.from_glib_suffix)
     }
+    fn generate_func_parameters(&self) -> String {
+        let mut param_str = String::with_capacity(100);
+        for (pos, &(ref par, out_as_returns)) in self.parameters.iter().enumerate() {
+            if pos > 0 { param_str.push_str(", ") }
+            let s = par;
+            if out_as_returns { param_str.push_str("&") }
+            param_str.push_str(&s);
+        }
+        param_str
+    }
+    fn get_outs(&self) -> Vec<&String> {
+        let outs: Vec<&String> = self.parameters.iter()
```

Comment:
Doesn't the compiler infer the correct `collect` from the function signature without the `outs` variable?